packages/pigeon/lib/ast.dart

```diff
@@ -82,39 +74,72 @@ class Api extends Node {
   }
 }
 
-/// Represents a field on a [Class].
-class Field extends Node {
-  /// Parametric constructor for [Field].
-  Field({
-    required this.name,
+/// An entity that represents a typed concept, like a [TypeDeclaration] or [NamedType].
+abstract class TypedEntity {
+  /// The datatype base name of the entity (ex 'Foo' to 'Foo<Bar>?').
+  String get dataType;
+
+  /// The type arguments to the entity.
+  List<TypeDeclaration>? get typeArguments;
+
+  /// True if the type is nullable.
+  bool get isNullable;
+}
+
+/// A specific instance of a type.
+class TypeDeclaration implements TypedEntity {
+  /// Constructor for [TypeDeclaration].
+  TypeDeclaration({
     required this.dataType,
     required this.isNullable,
     this.typeArguments,
-    this.offset,
   });
 
-  /// The name of the field.
+  @override
+  final String dataType;
+
+  @override
+  final List<TypeDeclaration>? typeArguments;
+
+  @override
+  final bool isNullable;
+
+  @override
+  String toString() {
+    return '(TypeDeclaration dataType:$dataType isNullable:$isNullable typeArguments:$typeArguments)';
+  }
+}
+
+/// Represents a named entity that has a type.
+class NamedType extends Node implements TypedEntity {
```

Comment:
Optional: a possible name that occurred to me last night is `NamedTypedEntity`. This addresses my confusion with the name (since it's not the type that is named here), and is completely consistent with `TypedEntity`.

I can live with `NamedType` if you hate that suggestion, but it would be much clearer to me as `NamedTypedEntity`.

Reply:
I don't hate it, i'm still on the fence about if we'll keep TypedEntity so I don't want to tie the names together.  It doesn't seem to be obviously better across all the criteria.